File: oled.ts
import { RIO } from 'rpi-io';
import SPI, { SpiDevice } from 'spi-device';
import { logger } from '../logger';
import { Canvas, CanvasRenderingContext2D } from 'skia-canvas';
import { sleep } from '../util/time';
import config from '../config';

export default class OLED {
  // Command constants from display datasheet
  // private static readonly ENABLE_GRAY_SCALE_TABLE = 0x00
  private static readonly SET_COLUMN_ADDRESS = 0x15;
  private static readonly WRITE_RAM = 0x5c;
  // private static readonly READ_RAM = 0x5D
  private static readonly SET_ROW_ADDRESS = 0x75;
  private static readonly SET_REMAP_DUAL_COM_LINE_MODE = 0xa0; // Re-map & Dual COM Line Mode
  private static readonly SET_DISPLAY_START_LINE = 0xa1;
  private static readonly SET_DISPLAY_OFFSET = 0xa2;
  // private static readonly SET_DISPLAY_MODE_ALL_OFF = 0xA4
  // private static readonly SET_DISPLAY_MODE_ALL_ON = 0xA5
  private static readonly SET_DISPLAY_MODE_NORMAL = 0xa6;
  // private static readonly SET_DISPLAY_MODE_INVERSE = 0xA7
  // private static readonly PARTIAL_DISPLAY_ENABLE = 0xA8
  private static readonly PARTIAL_DISPLAY_DISABLE = 0xa9;
  private static readonly SET_FUNCTION_SELECTION = 0xab;
  private static readonly DISPLAY_SLEEP_ON = 0xae;
  private static readonly DISPLAY_SLEEP_OFF = 0xaf;
  private static readonly SET_PHASE_LENGTH = 0xb1;
  private static readonly SET_FRONT_CLOCK_DIVIDER = 0xb3;
  private static readonly DISPLAY_ENHANCEMENT_A = 0xb4;
  private static readonly SET_GPIO = 0xb5;
  private static readonly SET_SECOND_PRECHARGE_PERIOD = 0xb6;
  // private static readonly SET_GRAY_SCALE_TABLE = 0xB8
  private static readonly SELECT_DEFAULT_LINEAR_GRAY_SCALE_TABLE = 0xb9;
  private static readonly SET_PRECHARGE_VOLTAGE = 0xbb;
  private static readonly SET_VCOMH_VOLTAGE = 0xbe;
  private static readonly SET_CONTRAST_CURRENT = 0xc1;
  private static readonly MASTER_CURRENT_CONTROL = 0xc7;
  private static readonly SET_MULTIPLEX_RATIO = 0xca;
  private static readonly DISPLAY_ENHANCEMENT_B = 0xd1;
  private static readonly SET_COMMAND_LOCK = 0xfd;

  // Options for controlling VSL selection
  private static readonly ENABLE_EXTERNAL_VSL = 0x00;
  // private static readonly ENABLE_INTERNAL_VSL = 0x02

  // Options for grayscale quality
  // private static readonly NORMAL_GRAYSCALE_QUALITY = 0xB0
  private static readonly ENHANCED_LOW_GRAY_SCALE_QUALITY = 0xf8;

  // Options for display enhancement b
  private static readonly RESERVED_ENHANCEMENT = 0x00;
  // private static readonly NORMAL_ENHANCEMENT = 0x02

  // Options for command lock
  // private static readonly COMMANDS_LOCK = 0x16
  private static readonly COMMANDS_UNLOCK = 0x12;

  // Column and row maximums
  // private static readonly COLUMN_ADDRESS = 0x77
  // private static readonly ROW_ADDRESS = 0x7F

  private spi: SpiDevice;
  private dcPin: RIO;
  private resetPin: RIO;
  public width: number;
  public height: number;
  private flip: boolean;
  private frameCanvas: Canvas;
  private frameCtx: CanvasRenderingContext2D;
  private frameBuf: Buffer;

  constructor(canvas: Canvas, flip = false) {
    this.frameCanvas = canvas;
    this.width = canvas.width;
    this.height = canvas.height;
    this.frameCtx = this.frameCanvas.getContext('2d');
    this.flip = flip;

    this.spi = SPI.openSync(
      config.get('oled.spiBus'),
      config.get('oled.spiDevice'),
      {
        maxSpeedHz: config.get('oled.spiSpeed'),
      }
    );

    this.dcPin = new RIO(config.get('oled.dataCommandPin'), 'output', {
      value: 0,
      bias: 'pull-up',
    });
    this.resetPin = new RIO(config.get('oled.resetPin'), 'output', {
      value: 1,
      bias: 'pull-up',
    });

    this.frameBuf = Buffer.alloc((this.width / 2) * this.height);
  }

  close() {
    this.writeCmd(OLED.DISPLAY_SLEEP_ON);
    this.spi.closeSync();
    this.dcPin.close();
    this.resetPin.close();
  }

  async init() {
    logger.debug('Display Initting');
    await this.reset();
    this.writeCmd(OLED.SET_COMMAND_LOCK, OLED.COMMANDS_UNLOCK);
    this.writeCmd(OLED.DISPLAY_SLEEP_ON);
    this.setAddress(0, 0, this.width / 4 - 1, this.height - 1);
    this.writeCmd(OLED.SET_FRONT_CLOCK_DIVIDER, 0x91); // B3
    this.writeCmd(OLED.SET_MULTIPLEX_RATIO, 0x3f);
    this.writeCmd(OLED.SET_DISPLAY_OFFSET, 0);
    this.writeCmd(OLED.SET_DISPLAY_START_LINE, 0);
    this.writeCmd(
      OLED.SET_REMAP_DUAL_COM_LINE_MODE,
      ///--------- Always 0
      //||/------- Always 0
      //|||/------ 0: Disable COM split, 1: Enable
      //||||/----- 0: Scan from beginning, 1: Scan from end
      //|||||/---- Always 0
      //||||||/--- 0: Normal, 1: Flip pixel nibbles
      //|||||||/-- 0: Normal, 1: Flip columns left-to-right
      //||||||||/- 0: Scan addresses horizontally, 1: Vertically
      0b00010100,

      ///--------- Always 0
      //||/------- Always 0
      //|||/------ Always 0
      //||||/----- 0: Disable Dual COM mode, 1: Enable
      //|||||/---- Always 0
      //||||||/--- Always 0
      //|||||||/-- Always 1
      //||||||||/- Always 1
      0x00010011
    ); // A0
    this.writeCmd(OLED.SET_GPIO, 0); // B5
    this.writeCmd(OLED.SET_FUNCTION_SELECTION, 1); // AB
    this.writeCmd(
      OLED.DISPLAY_ENHANCEMENT_A,
      OLED.ENABLE_EXTERNAL_VSL | 0xa0,
      OLED.ENHANCED_LOW_GRAY_SCALE_QUALITY | 0x05
    );
    this.writeCmd(OLED.SET_CONTRAST_CURRENT, 0xff); // C1
    this.writeCmd(OLED.MASTER_CURRENT_CONTROL, 0xf);
    this.writeCmd(OLED.SELECT_DEFAULT_LINEAR_GRAY_SCALE_TABLE);
    this.writeCmd(OLED.SET_PHASE_LENGTH, 0xe2); // B1
    this.writeCmd(OLED.SET_SECOND_PRECHARGE_PERIOD, 0x8); // B6
    this.writeCmd(
      OLED.DISPLAY_ENHANCEMENT_B,
      OLED.RESERVED_ENHANCEMENT | 0xa2,
      0x20
    );
    this.writeCmd(OLED.SET_PRECHARGE_VOLTAGE, 0x1f); // BB
    this.writeCmd(OLED.SET_VCOMH_VOLTAGE, 0x7); // BE
    this.writeCmd(OLED.SET_DISPLAY_MODE_NORMAL);
    this.writeCmd(OLED.PARTIAL_DISPLAY_DISABLE);
    this.writeCmd(OLED.DISPLAY_SLEEP_OFF);
    logger.debug('Display Initted');
  }

  setBrightness(byte: number) {
    this.writeCmd(OLED.SET_CONTRAST_CURRENT, byte & 0xff);
    // this.writeCmd(OLED.MASTER_CURRENT_CONTROL, 15)
  }

  async toPng(): Promise<Buffer> {
    return this.frameCanvas.toBuffer('png');
  }

  // -------------------------------------
  // Private
  // -------------------------------------

  public async reset() {
    this.resetPin.write(0);
    await sleep(200);
    this.resetPin.write(1);
    await sleep(200);
  }

  private spiWrite(send: Buffer) {
    this.spi.transferSync([
      {
        sendBuffer: send,
        byteLength: send.byteLength,
      },
    ]);
  }

  private writeData(data: Buffer) {
    this.dcPin.write(1);
    this.spiWrite(data);
  }

  private writeCmd(cmd: number, ...data: number[]) {
    this.dcPin.write(0);

    if (data?.length) {
      this.spiWrite(Buffer.from([cmd]));
      return this.writeData(Buffer.from(data));
    } else {
      return this.spiWrite(Buffer.from([cmd]));
    }
  }

  private setColumnAddress(start: number, end: number) {
    this.writeCmd(OLED.SET_COLUMN_ADDRESS, start, end);
  }

  private setRowAddress(start: number, end: number) {
    this.writeCmd(OLED.SET_ROW_ADDRESS, start, end);
  }

  private setAddress(
    x0: number,
    y0: number,
    x1: number,
    y1: number,
    offset = 0x1c
  ) {
    this.setRowAddress(y0, y1);
    this.setColumnAddress(x0 + offset, x1 + offset);
    this.writeCmd(OLED.WRITE_RAM);
  }

  // 2. Convert to grayscale
  // 3. Write out SPI off-screen
  // 4. Move screen to show new frame
  public async blit(num: number) {
    // 1. Save current frame
    const image = this.frameCtx.getImageData(0, 0, this.width, this.height);
    const buf = this.frameBuf;

    // 2. Convert to 4-bit gray in frameBuff
    let framePtr = this.flip ? (this.width / 2) * this.height - 1 : 0;
    const inc = this.flip ? -1 : 1;
    const px = image.data;
    let limit = px.byteLength;

    for (let p = 0; p < limit; p += 8) {
      buf[framePtr] =
        toGray(px[p], px[p + 1], px[p + 2], px[p + 3]) |
        (toGray(px[p + 4], px[p + 5], px[p + 6], px[p + 7]) << 4);
      framePtr += inc;
    }

    const yStart = num % 2 === 1 ? this.height : 0;
    const displayOffset = num % 2 === 1 ? this.height : 0;
    const step = 4096;

    // 3. Double buffering… Move to an area that's not on screen right now
    this.setAddress(0, yStart, this.width / 4 - 1, yStart + this.height - 1);

    // Write pixels there
    limit = buf.byteLength / step;
    for (let i = 0; i < limit; i++) {
      this.writeData(buf.subarray(step * i, step * (i + 1)));
    }

    // 4. Move the active screen to the new pixels
    this.writeCmd(OLED.SET_DISPLAY_START_LINE, displayOffset);
  }
}

function toGray(r: number, g: number, b: number, a: number) {
  return Math.round(((0.3 * r + 0.59 * g + 0.11 * b) * a) / 255) >> 4;
}

File: server/hardware/oled/oled.go
// SSD1327-based 4-bit grayscale OLED display over SPI.
// Ported from oled.ts; tested against a 256×64 panel.
//
// Wiring:
//
//	SPI MOSI/CLK/CS → standard SPI bus pins
//	DC pin          → GPIO output (low = command, high = data)
//	Reset pin       → GPIO output (low = reset, high = run)
package oled

import (
	"image"
	"image/color"
	"image/draw"
	"math"
	"time"

	"github.com/warthog618/go-gpiocdev"
	"periph.io/x/conn/v3/physic"
	"periph.io/x/conn/v3/spi"
	"periph.io/x/conn/v3/spi/spireg"
	"periph.io/x/host/v3"
)

// Command constants from the SSD1327 datasheet.
const (
	setColumnAddress                  = 0x15
	writeRAM                          = 0x5c
	setRowAddress                     = 0x75
	setRemapDualComLineMode           = 0xa0
	setDisplayStartLine               = 0xa1
	setDisplayOffset                  = 0xa2
	setDisplayModeNormal              = 0xa6
	partialDisplayDisable             = 0xa9
	setFunctionSelection              = 0xab
	displaySleepOn                    = 0xae
	displaySleepOff                   = 0xaf
	setPhaseLength                    = 0xb1
	setFrontClockDivider              = 0xb3
	displayEnhancementA               = 0xb4
	setGPIO                           = 0xb5
	setSecondPrechargePeriod          = 0xb6
	selectDefaultLinearGrayScaleTable = 0xb9
	setPrechargeVoltage               = 0xbb
	setVCOMHVoltage                   = 0xbe
	setContrastCurrent                = 0xc1
	masterCurrentControl              = 0xc7
	setMultiplexRatio                 = 0xca
	displayEnhancementB               = 0xd1
	setCommandLock                    = 0xfd

	enableExternalVSL           = 0x00
	enhancedLowGrayScaleQuality = 0xf8
	reservedEnhancement         = 0x00
	commandsUnlock              = 0x12

	// columnOffset is the hardware column offset for this panel.
	columnOffset = 0x1c
)

// Config holds the hardware configuration for the OLED.
type Config struct {
	// SPIPort is the spidev path, e.g. "/dev/spidev0.0".
	SPIPort string
	// SPISpeed is the SPI clock frequency.
	SPISpeed physic.Frequency
	// GPIOChip is the gpiochip device, e.g. "gpiochip0".
	GPIOChip string
	// DCPin is the BCM GPIO line number for the data/command pin.
	DCPin int
	// ResetPin is the BCM GPIO line number for the reset pin.
	ResetPin int
	// Flip reverses the frame buffer before writing (180° rotation).
	Flip bool
}

// OLED drives a 4-bit grayscale SSD1327 display over SPI.
type OLED struct {
	cfg      Config
	width    int
	height   int
	spiPort  spi.PortCloser
	spiConn  spi.Conn
	dcLine   *gpiocdev.Line
	rstLine  *gpiocdev.Line
	frameBuf []byte
	frameNum int64
}

// New opens the SPI bus and GPIO lines, then initialises the display.
// The caller supplies width and height in pixels.
func New(cfg Config, width, height int) (*OLED, error) {
	if _, err := host.Init(); err != nil {
		return nil, err
	}

	port, err := spireg.Open(cfg.SPIPort)
	if err != nil {
		return nil, err
	}

	conn, err := port.Connect(cfg.SPISpeed, spi.Mode0, 8)
	if err != nil {
		port.Close()
		return nil, err
	}

	chip := cfg.GPIOChip
	if chip == "" {
		chip = "gpiochip0"
	}

	dcLine, err := gpiocdev.RequestLine(chip, cfg.DCPin,
		gpiocdev.AsOutput(0),
		gpiocdev.WithPullUp,
	)
	if err != nil {
		port.Close()
		return nil, err
	}

	rstLine, err := gpiocdev.RequestLine(chip, cfg.ResetPin,
		gpiocdev.AsOutput(1),
		gpiocdev.WithPullUp,
	)
	if err != nil {
		dcLine.Close()
		port.Close()
		return nil, err
	}

	o := &OLED{
		cfg:      cfg,
		width:    width,
		height:   height,
		spiPort:  port,
		spiConn:  conn,
		dcLine:   dcLine,
		rstLine:  rstLine,
		frameBuf: make([]byte, (width/2)*height),
	}

	if err := o.Init(); err != nil {
		o.Close()
		return nil, err
	}

	return o, nil
}

// Close puts the display to sleep and releases all hardware resources.
func (o *OLED) Close() {
	o.writeCmd(displaySleepOn)
	o.spiPort.Close()
	o.dcLine.Close()
	o.rstLine.Close()
}

// Init resets the display and sends the full initialisation sequence.
func (o *OLED) Init() error {
	if err := o.Reset(); err != nil {
		return err
	}

	o.writeCmd(setCommandLock, commandsUnlock)
	o.writeCmd(displaySleepOn)
	o.setAddress(0, 0, o.width/4-1, o.height-1)
	o.writeCmd(setFrontClockDivider, 0x91)
	o.writeCmd(setMultiplexRatio, 0x3f)
	o.writeCmd(setDisplayOffset, 0)
	o.writeCmd(setDisplayStartLine, 0)
	o.writeCmd(setRemapDualComLineMode,
		0b00010100,
		0b00010011,
	)
	o.writeCmd(setGPIO, 0)
	o.writeCmd(setFunctionSelection, 1)
	o.writeCmd(displayEnhancementA,
		enableExternalVSL|0xa0,
		enhancedLowGrayScaleQuality|0x05,
	)
	o.writeCmd(setContrastCurrent, 0xff)
	o.writeCmd(masterCurrentControl, 0xf)
	o.writeCmd(selectDefaultLinearGrayScaleTable)
	o.writeCmd(setPhaseLength, 0xe2)
	o.writeCmd(setSecondPrechargePeriod, 0x8)
	o.writeCmd(displayEnhancementB,
		reservedEnhancement|0xa2,
		0x20,
	)
	o.writeCmd(setPrechargeVoltage, 0x1f)
	o.writeCmd(setVCOMHVoltage, 0x7)
	o.writeCmd(setDisplayModeNormal)
	o.writeCmd(partialDisplayDisable)
	o.writeCmd(displaySleepOff)

	return nil
}

// SetBrightness sets the display contrast (0–255).
func (o *OLED) SetBrightness(b byte) {
	o.writeCmd(setContrastCurrent, b)
}

// Reset pulses the reset pin low for 200 ms then releases it.
func (o *OLED) Reset() error {
	if err := o.rstLine.SetValue(0); err != nil {
		return err
	}
	time.Sleep(200 * time.Millisecond)
	if err := o.rstLine.SetValue(1); err != nil {
		return err
	}
	time.Sleep(200 * time.Millisecond)

	black := image.NewRGBA(image.Rect(0, 0, o.width, o.height))
	draw.Draw(black, black.Bounds(), &image.Uniform{color.RGBA{0, 0, 0, 255}}, image.Point{}, draw.Src)

	o.Blit(black)
	o.Blit(black)

	return nil
}

// Blit converts img to 4-bit grayscale and writes it to the display using
// double buffering. num alternates between frames (even / odd) to avoid
// tearing while the panel scrolls to the new buffer.
func (o *OLED) Blit(img image.Image) {
	bounds := img.Bounds()
	buf := o.frameBuf

	framePtr := 0
	inc := 1
	if o.cfg.Flip {
		framePtr = len(buf) - 1
		inc = -1
	}

	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x += 2 {
			hi := toGray(img.At(x, y))
			lo := toGray(img.At(x+1, y))
			if o.cfg.Flip {
				buf[framePtr] = hi | (lo << 4)
			} else {
				buf[framePtr] = lo | (hi << 4)
			}
			framePtr += inc
		}
	}

	yStart := 0
	displayOffset := 0
	if o.frameNum%2 == 1 {
		yStart = o.height
		displayOffset = o.height
	}
	o.frameNum++

	// Write pixels into the off-screen buffer area.
	o.setAddress(0, yStart, o.width/4-1, yStart+o.height-1)

	const step = 4096
	for i := 0; i < len(buf); i += step {
		end := min(i+step, len(buf))
		o.writeData(buf[i:end])
	}

	// Flip the display start line to reveal the new frame.
	o.writeCmd(setDisplayStartLine, byte(displayOffset))
}

// Width returns the display width in pixels.
func (o *OLED) Width() int { return o.width }

// Height returns the display height in pixels.
func (o *OLED) Height() int { return o.height }

// -------------------------------------------------------------------------
// Private helpers
// -------------------------------------------------------------------------

func (o *OLED) spiWrite(data []byte) {
	_ = o.spiConn.Tx(data, nil)
}

func (o *OLED) writeData(data []byte) {
	_ = o.dcLine.SetValue(1)
	o.spiWrite(data)
}

func (o *OLED) writeCmd(cmd byte, data ...byte) {
	_ = o.dcLine.SetValue(0)
	o.spiWrite([]byte{cmd})
	if len(data) > 0 {
		o.writeData(data)
	}
}

func (o *OLED) setColumnAddress(start, end int) {
	o.writeCmd(setColumnAddress, byte(start), byte(end))
}

func (o *OLED) setRowAddress(start, end int) {
	o.writeCmd(setRowAddress, byte(start), byte(end))
}

func (o *OLED) setAddress(x0, y0, x1, y1 int) {
	o.setRowAddress(y0, y1)
	o.setColumnAddress(x0+columnOffset, x1+columnOffset)
	o.writeCmd(writeRAM)
}

// toGray converts a pixel to a 4-bit (0–15) grayscale value using the same
// weighted luminance formula as the original TypeScript implementation.
// It respects alpha by premultiplying before quantising.
func toGray(c interface{ RGBA() (r, g, b, a uint32) }) byte {
	r, g, b, a := c.RGBA()
	if a == 0 {
		return 0
	}
	// image.Color returns 16-bit channels; scale to 0–255.
	rf := float64(r>>8) * 0.30
	gf := float64(g>>8) * 0.59
	bf := float64(b>>8) * 0.11
	af := float64(a>>8) / 255.0
	gray := math.Round((rf+gf+bf)*af) / 16.0
	v := byte(gray)
	if v > 15 {
		v = 15
	}
	return v
}
